File: topologicalSort.java
import java.util.*;

public class topologicalSort {
    static class Edge {
        int src;
        int nbr;
        int wt;

        Edge(int src, int nbr, int wt) {
            this.src = src;
            this.nbr = nbr;
            this.wt = wt;
        }
    }

    public static void topologicalSortSolution(ArrayList<Edge>[]graph,int src,boolean[]visited,Stack<Integer>st)
    {
        visited[src] = true;
        for(Edge ele:graph[src])
        {
            if(visited[ele.nbr]==false)
            {
                topologicalSortSolution(graph, ele.nbr, visited, st);
            }
        }
        st.push(src);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        
        ArrayList<Edge>[] graph = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            graph[i] = new ArrayList<>();
        }
        for (int i = 0; i < n; i++) {
            int src = sc.nextInt();
            int nbr = sc.nextInt();
            // int wt = sc.nextInt();
            graph[src].add(new Edge(src, nbr, 0));
            
        }
        Stack<Integer>st = new Stack<>();
        boolean []visited = new boolean[n];
        int src = sc.nextInt();
        // ArrayList<ArrayList<Integer>> l= new ArrayList<>();
        topologicalSortSolution(graph, src, visited, st);
        while (st.empty()==false) {
            System.out.print(st.pop()+" ");
        }
        System.out.println();
    } 
}
/*
 * basic graph input
 * 7
 * 8
 * 0 1 10
 * 1 2 10
 * 2 3 10
 * 0 3 10
 * 3 4 10
 * 4 5 10
 * 5 6 10
 * 4 6 10
 * 0
 * 6
 */
/*
 * 7
 * 8
 * 0 1 10
 * 0 3 40
 * 1 2 10
 * 2 3 10
 * 3 4 2
 * 4 5 3
 * 4 6 8
 * 5 6 3
 * 0
 */
/*
 * 7
 * 5
 * 0 1 10
 * 2 3 10
 * 4 5 10
 * 5 6 10
 * 4 6 10
 */

 /*abstract



 7
7
0 1 10
1 2 10 
2 3 10
0 3 10
4 5 10
5 6 10
4 6 10


*/

File: TopologicalSort.cpp
#include <bits/stdc++.h>
using namespace std;

void topologicalSort(vector<int> adj[], vector<bool> &vis, int src, stack<int> &st)
{
    vis[src] = true;
    for (auto ele : adj[src])
    {
        if (vis[ele] == false)
        {
            topologicalSort(adj, vis, ele, st);
        }
    }
    st.push(src);
}

int main()
{
    int n;
    cin >> n;
    int m;
    cin >> m;
    vector<int> adj[n];
    for (int i = 0; i < m; i++)
    {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        // adj[v].push_back(u);
    }

    vector<bool> vis(n, false);
    stack<int> st;
    vector<int> ans;
    for (int i = 0; i < n; i++)
    {
        if (vis[i] == false)
        {
            topologicalSort(adj, vis, i, st);
        }
    }
    while (!st.empty())
    {
        ans.push_back(st.top());
        st.pop();
    }
    for (auto ele : ans)
    {
        cout << ele << " ";
    }
    cout << endl;

    return 0;
}

//? OUTPUT GOES HERE
/**
 * @brief
 *
6
6
5 0
5 2
4 0
2 3
1 3
4 1
5 4 2 1 3 0
 *
 */

//! INPUT OF THE HACKER EARTH

/*

5 6
1 2
1 3
2 3
2 4
3 4
3 5

*/

//! PEPCODING SAME OLD BANGER

/*


7
8
0 1
0 3
1 2
2 3
4 3
4 5
4 6
5 6
4 5 6 0 1 2 3 

*/